File: app/src/main/java/jp/hazuki/yuzubrowser/gesture/multiFinger/data/MultiFingerGestureManager.java
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.gesture.multiFinger.data;

import android.content.Context;

import com.fasterxml.jackson.core.JsonEncoding;
import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import jp.hazuki.yuzubrowser.action.Action;
import jp.hazuki.yuzubrowser.utils.ArrayUtils;

public class MultiFingerGestureManager {
    private static final String FILENAME = "multiFingerGes_1.dat";
    private final File JSON_FILE;
    private List<MultiFingerGestureItem> gestureItems;

    public MultiFingerGestureManager(Context context) {
        gestureItems = new ArrayList<>();
        JSON_FILE = context.getFileStreamPath(FILENAME);
        load();
    }

    public void add(MultiFingerGestureItem item) {
        gestureItems.add(item);
        save();
    }

    public void add(int index, MultiFingerGestureItem item) {
        gestureItems.add(index, item);
        save();
    }

    public void set(int index, MultiFingerGestureItem item) {
        gestureItems.set(index, item);
        save();
    }

    public MultiFingerGestureItem remove(int index) {
        MultiFingerGestureItem item = gestureItems.remove(index);
        save();
        return item;
    }

    public void remove(MultiFingerGestureItem item) {
        if (gestureItems.remove(item))
            save();
    }

    public void move(int from, int to) {
        ArrayUtils.move(gestureItems, from, to);
        save();
    }

    public int indexOf(MultiFingerGestureItem item) {
        return gestureItems.indexOf(item);
    }

    public List<MultiFingerGestureItem> getGestureItems() {
        return gestureItems;
    }

    private static final String JSON_FINGERS = "0";
    private static final String JSON_TRACES = "1";
    private static final String JSON_ACTION = "2";

    private void load() {
        gestureItems.clear();
        if (JSON_FILE == null || !JSON_FILE.exists() || JSON_FILE.isDirectory()) return;
        JsonFactory factory = new JsonFactory();
        try (JsonParser parser = factory.createParser(JSON_FILE)) {
            if (parser.nextToken() == JsonToken.START_ARRAY) {
                while (parser.nextToken() != JsonToken.END_ARRAY) {
                    if (parser.getCurrentToken() == JsonToken.START_OBJECT) {
                        MultiFingerGestureItem item = new MultiFingerGestureItem();
                        while (parser.nextToken() != JsonToken.END_OBJECT) {
                            String name = parser.getCurrentName();
                            if (name != null) {
                                switch (name) {
                                    case JSON_FINGERS:
                                        parser.nextToken();
                                        item.setFingers(parser.getIntValue());
                                        break;
                                    case JSON_TRACES:
                                        if (parser.nextToken() == JsonToken.START_ARRAY) {
                                            while (parser.nextToken() != JsonToken.END_ARRAY) {
                                                item.addTrace(parser.getIntValue());
                                            }
                                        } else {
                                            parser.skipChildren();
                                        }
                                        break;
                                    case JSON_ACTION:
                                        Action action = new Action();
                                        action.loadAction(parser);
                                        item.setAction(action);
                                        break;
                                    default:
                                        parser.skipChildren();
                                        break;
                                }
                            }
                        }
                        gestureItems.add(item);
                    } else {
                        parser.skipChildren();
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void save() {
        JsonFactory jsonFactory = new JsonFactory();
        try (JsonGenerator generator = jsonFactory.createGenerator(JSON_FILE, JsonEncoding.UTF8)) {
            generator.writeStartArray();
            for (MultiFingerGestureItem item : gestureItems) {
                generator.writeStartObject();

                generator.writeNumberField(JSON_FINGERS, item.getFingers());

                // finger actions
                generator.writeFieldName(JSON_TRACES);
                generator.writeStartArray();
                for (Integer i : item.getTraces()) {
                    generator.writeNumber(i);
                }
                generator.writeEndArray();

                // browser action
                generator.writeFieldName(JSON_ACTION);
                item.getAction().writeAction(generator);

                generator.writeEndObject();
            }
            generator.writeEndArray();
            generator.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/gesture/multiFinger/data/MultiFingerGestureManager.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.gesture.multiFinger.data

import android.content.Context
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.core.utility.utils.ArrayUtils
import jp.hazuki.yuzubrowser.legacy.action.Action
import okio.Okio
import okio.buffer
import okio.sink
import okio.source
import java.io.IOException
import java.util.*

class MultiFingerGestureManager(context: Context) {
    private val jsonFile = context.getFileStreamPath(FILENAME)
    val gestureItems: MutableList<MultiFingerGestureItem> = ArrayList()

    init {
        load()
    }

    fun add(item: MultiFingerGestureItem) {
        gestureItems.add(item)
        save()
    }

    fun add(index: Int, item: MultiFingerGestureItem) {
        gestureItems.add(index, item)
        save()
    }

    operator fun set(index: Int, item: MultiFingerGestureItem) {
        gestureItems[index] = item
        save()
    }

    fun remove(index: Int): MultiFingerGestureItem {
        val item = gestureItems.removeAt(index)
        save()
        return item
    }

    fun remove(item: MultiFingerGestureItem) {
        if (gestureItems.remove(item))
            save()
    }

    fun move(from: Int, to: Int) {
        ArrayUtils.move(gestureItems, from, to)
        save()
    }

    fun indexOf(item: MultiFingerGestureItem): Int {
        return gestureItems.indexOf(item)
    }

    private fun load() {
        gestureItems.clear()
        if (jsonFile == null || !jsonFile.exists() || jsonFile.isDirectory) return
        try {
            JsonReader.of(jsonFile.source().buffer()).use { reader ->
                if (reader.peek() == JsonReader.Token.BEGIN_ARRAY) {
                    reader.beginArray()
                    while (reader.hasNext()) {
                        if (reader.peek() == JsonReader.Token.BEGIN_OBJECT) {
                            reader.beginObject()
                            val item = MultiFingerGestureItem()
                            while (reader.hasNext()) {
                                when (reader.nextName()) {
                                    JSON_FINGERS -> {
                                        item.fingers = reader.nextInt()
                                    }
                                    JSON_TRACES -> if (reader.peek() == JsonReader.Token.BEGIN_ARRAY) {
                                        reader.beginArray()
                                        while (reader.hasNext()) {
                                            item.addTrace(reader.nextInt())
                                        }
                                        reader.endArray()
                                    } else {
                                        reader.skipValue()
                                    }
                                    JSON_ACTION -> {
                                        val action = Action()
                                        action.loadAction(reader)
                                        item.action = action
                                    }
                                    else -> reader.skipValue()
                                }
                            }
                            gestureItems.add(item)
                            reader.endObject()
                        } else {
                            reader.skipValue()
                        }
                    }
                    reader.endArray()
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun save() {
        try {
            JsonWriter.of(jsonFile.sink().buffer()).use { writer ->
                writer.beginArray()
                gestureItems.forEach { item ->
                    writer.beginObject()

                    writer.name(JSON_FINGERS)
                    writer.value(item.fingers)

                    // finger actions
                    writer.name(JSON_TRACES)
                    writer.beginArray()
                    item.traces.forEach { writer.value(it) }
                    writer.endArray()

                    // browser action
                    writer.name(JSON_ACTION)
                    item.action.writeAction(writer)

                    writer.endObject()
                }
                writer.endArray()
                writer.flush()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    companion object {
        private const val FILENAME = "multiFingerGes_1.dat"

        private const val JSON_FINGERS = "0"
        private const val JSON_TRACES = "1"
        private const val JSON_ACTION = "2"
    }
}
